### PR TITLE
Refactor ActiveLedgerState to use F-bounded polymorphism

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
Minor refactoring opportunity I came across.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/ActiveLedgerState.scala [hunk 1]
```diff
@@ -50,16 +50,16 @@ trait ActiveLedgerState[+Self] { this: ActiveLedgerState[Self] =>
   def lookupContractByKey(key: GlobalKey): Option[AbsoluteContractId]
 
   /** Called when a new contract is created */
-  def addContract(c: ActiveContract, keyO: Option[GlobalKey]): Self
+  def addContract(c: ActiveContract, keyO: Option[GlobalKey]): ALS
```
Comment (SamirTalwar-DA) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/store/ActiveLedgerState.scala]: Couldn't you just use `this.type` as the return values, and get rid of the generic-ness altogether?

Comment (stefanobaghino-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/ActiveLedgerState.scala]: No. `this.type` is a singleton type and forces to return the very same instance on which the call was made, which is not necessarily the case in implementations (which may be immutable and thus need to return a copy to properly implement the `trait`, behavior-wise). With F-bounded polymorphism we're just restricting to the implementation type. It's more restrictive at compile time than the existing use of covariant return types and allows us to drop the usage of the implicit evidence.

Comment (SamirTalwar-DA) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/ActiveLedgerState.scala]: I'm convinced. 🙂

**Comment by gerolf-da** [comment]
I think we can actually remove the type parameters from `ActiveLedgerState` and `ActiveLedgerStateManager`. `Self` or `ALS` is only used in the return type of methods in `ActiveLedgerState`, and implementations can just override the return type of the methods to return the more specific subtype.

**Comment by SamirTalwar-DA** [comment]
@gerolf-da: I think the return value of `addTransaction` needs to be the subtype, because the return value is then deconstructed.

**Comment by gerolf-da** [comment]
Having said that, let's also not make this a time sink, therefore 👍 .

**Pull request merged**